clients/arweave_ipfs_s3.go

```diff
@@ -2,52 +2,138 @@ package clients
 
 import (
 	"fmt"
+	"io"
 	"net/http"
 	"net/url"
 	"strings"
 	"time"
 
-	"github.com/hashicorp/go-retryablehttp"
+	"github.com/cenkalti/backoff/v4"
+	"github.com/livepeer/catalyst-api/config"
+	"github.com/livepeer/catalyst-api/log"
 )
 
+const SCHEME_IPFS = "ipfs"
+const SCHEME_ARWEAVE = "ar"
+
 const MAX_COPY_DURATION = 20 * time.Minute
 
-var defaultRetryableHttpClient = newDefaultRetryableHttpClient()
+func CopyDStorageToS3(url, s3URL string, requestID string) error {
+	content, err := DownloadDStorageFromGatewayList(url, requestID)
+	if err != nil {
+		return err
+	}
 
-func newDefaultRetryableHttpClient() *http.Client {
-	client := retryablehttp.NewClient()
-	client.RetryMax = 2                          // Retry a maximum of this+1 times
-	client.RetryWaitMin = 200 * time.Millisecond // Wait at least this long between retries
-	client.RetryWaitMax = 1 * time.Second        // Wait at most this long between retries (exponential backoff)
-	client.HTTPClient = &http.Client{
-		// Give up on requests that take more than this long - the file is probably too big for us to process locally if it takes this long
-		// or something else has gone wrong and the request is hanging
-		Timeout: MAX_COPY_DURATION,
+	err = UploadToOSURL(s3URL, "", content, MAX_COPY_DURATION)
+	if err != nil {
+		return err
 	}
 
-	return client.StandardClient()
+	return nil
 }
 
-func CopyArweaveToS3(arweaveURL, s3URL string) error {
-	resp, err := defaultRetryableHttpClient.Get(arweaveURL)
+func DownloadDStorageFromGatewayList(u string, requestID string) (io.ReadCloser, error) {
+	var err error
+	var gateways []*url.URL
+	dStorageURL, err := url.Parse(u)
 	if err != nil {
-		return fmt.Errorf("error fetching Arweave or IPFS URL: %s", err)
+		return nil, fmt.Errorf("invalid URL: %w", err)
 	}
-	defer resp.Body.Close()
 
-	err = UploadToOSURL(s3URL, "", resp.Body, MAX_COPY_DURATION)
+	var resourceID string
+
+	if dStorageURL.Scheme == SCHEME_ARWEAVE {
+		gateways = config.ImportArweaveGatewayURLs
+		resourceID = dStorageURL.Host
+	} else if dStorageURL.Scheme == SCHEME_IPFS {
+		gateways = config.ImportIPFSGatewayURLs
+		resourceID = dStorageURL.Host
+	} else {
+		var gateway, dStorageType string
+		resourceID, gateway, dStorageType = parseDStorageGatewayURL(dStorageURL)
+		if dStorageType == "" {
+			return nil, fmt.Errorf("unsupported dStorage resource %s", dStorageURL.Scheme)
+		}
+
+		gatewayURL, err := url.Parse(gateway)
+		if err != nil {
+			return nil, fmt.Errorf("invalid URL: %w", err)
+		}
+
+		gateways = append(gateways, gatewayURL)
```

Comment:
**nit:** I was confused a bit thinking there would be something before the user-provided gateway. Just a nit to be clearer here:
```suggestion
		gateways = []*url.URL{gatewayURL}
```